### PR TITLE
Fscloud vsi

### PR DESCRIPTION
**Pull request by rajatagarwal-ibm**
### Description

FSCloud implementation for VSI

#### Changes that affect the core Terraform module or submodules
- [ ] Bug fix
- [x] New feature
- [ ] Dependency update

#### Changes that don't affect the core Terraform module or submodules
- [ ] Examples or tests (addition or updates of examples or tests)
- [ ] Documentation update
- [ ] CI-related update (pipeline, etc.)
- [ ] Other

#### Release required?
Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning).

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- Use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author. The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).
- Merge by using "Squash and merge".

### CODE REVIEW

File: profiles/fscloud/variables.tf [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+# Account Variables
+##############################################################################
+
+variable "resource_group_id" {
+  description = "id of resource group to create VPC"
+  type        = string
+}
+
+variable "prefix" {
+  description = "The IBM Cloud platform API key needed to deploy IAM enabled resources"
```
Comment (ocofaigh) [inline profiles/fscloud/variables.tf:11]: seems to be wrong descrption

File: examples/fscloud/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# End to end example with default values
+
+An end to end example using the module's default variable values.
```
Comment (ocofaigh) [inline examples/fscloud/README.md:3]: this readme needs to be updated

File: examples/default/variables.tf [hunk 3]
```diff
@@ -19,7 +19,7 @@ variable "region" {
 variable "prefix" {
   description = "The prefix that you would like to append to your resources"
   type        = string
-  default     = "test-landing-zone-vsi"
+  default     = "rajat-vsi"
```
Comment (ocofaigh) [inline examples/default/variables.tf:22]: dont use `rajat` in the prefix default value

File: variables.tf [hunk 4]
```diff
@@ -362,4 +362,10 @@ variable "secondary_allow_ip_spoofing" {
   default     = false
 }
 
+variable "existing_kms_instance_guid" {
+  description = "The GUID of the Hyper Protect Crypto Services instance in which the key specified in var.kms_key_crn is coming from."
```
Comment (ocofaigh) [inline variables.tf:366]: there is no variable called `kms_key_crn` in this module?

**Review by ocofaigh — COMMENTED** [review]
I started to review, but perhaps you are still working on this?

**Comment by rajatagarwal-ibm** [comment]
@ocofaigh yeah the work is still in progress... its much of a draft version... i was focusing on some other stuff today so will be back on this tomorrow.

File: examples/default/variables.tf (hunk 3 above)
Comment (rajatagarwal-ibm) [reply examples/default/variables.tf:22]: yeah was running tests locally.

File: variables.tf (hunk 4 above)
Comment (rajatagarwal-ibm) [reply variables.tf:366]: Yes, keys are passed in `boot_volume_encryption_key`

File: tests/pr_test.go [hunk 5]
```diff
@@ -66,3 +69,23 @@ func TestRunUpgradeBasicExample(t *testing.T) {
 		assert.NotNil(t, output, "Expected some output")
 	}
 }
+
```
Comment (ocofaigh) [inline tests/pr_test.go:72]: should the basic example now move to other_test.go like we did for all ICD modules? I think all the features will now be covered by fscloud complete example

File: storage.tf [hunk 6]
```diff
@@ -19,7 +19,7 @@ locals {
           capacity       = volume.capacity
           vsi_name       = "${var.prefix}-${(subnet) * (var.vsi_per_subnet) + count + 1}"
           iops           = volume.iops
-          encryption_key = volume.encryption_key
+          encryption_key = var.existing_kms_instance_guid != null ? var.boot_volume_encryption_key : volume.encryption_key
```
Comment (ocofaigh) [inline storage.tf:22]: `existing_kms_instance_guid` is not the right variable to use here. `existing_kms_instance_guid` is only needed if you want to create the auth policy. In ICS modules we use a variable called `kms_encryption_enabled`. Does anything like this exist in this module already? If not, I think we need it

File: main.tf [hunk 7]
```diff
@@ -64,6 +63,17 @@ locals {
 # Create Virtual Servers
 ##############################################################################
 
+resource "ibm_iam_authorization_policy" "block_storage_policy" {
+  count               = var.existing_kms_instance_guid == null || var.skip_iam_authorization_policy ? 0 : 1
```
Comment (ocofaigh) [inline main.tf:67]: `existing_kms_instance_guid` should not be used here.

File: profiles/fscloud/README.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
+# Profile for IBM Cloud Framework for Financial Services
+
+This code is a version of the [parent root module](../../) that includes a default configuration that complies with the relevant controls from the [IBM Cloud Framework for Financial Services](https://cloud.ibm.com/docs/framework-financial-services?topic=framework-financial-services-about). See the [Example for IBM Cloud Framework for Financial Services](/examples/fscloud/) for logic that uses this module.
+
+The default values in this profile were scanned by [IBM Code Risk Analyzer (CRA)](https://cloud.ibm.com/docs/code-risk-analyzer-cli-plugin?topic=code-risk-analyzer-cli-plugin-cra-cli-plugin#terraform-command) for compliance with the IBM Cloud Framework for Financial Services profile that is specified by the IBM Security and Compliance Center. The scan passed for all applicable goals with one exception:
+
+> rule-beb7b289-706b-4dc0-b01d-b1d15d4331e3: Check whether Databases for PostgreSQL network access is restricted to a specific IP range.
```
Comment (ocofaigh) [inline profiles/fscloud/README.md:7]: This seems a copy and paste from postgres. There are no failing rules right?

**Review by ocofaigh — CHANGES_REQUESTED** [review]
see comments

File: variables.tf [hunk 9]
```diff
@@ -184,6 +191,18 @@ variable "security_group_ids" {
   }
 }
 
+variable "kms_encryption_enabled" {
+  type        = bool
+  description = "Set this to true to control the encryption keys used to encrypt the data that you store in IBM Cloud® Databases. If set to false, the data is encrypted by using randomly generated keys. For more info on HPCS integration, see https://cloud.ibm.com/docs/cloud-databases?topic=cloud-databases-hpcs"
```
Comment (ocofaigh) [inline variables.tf:196]: This is a link to ICD - you should find a doc on boot drive encryption instead

Comment (ocofaigh) [reply variables.tf:196]: this probably https://cloud.ibm.com/docs/vpc?topic=vpc-creating-instances-byok

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**